File: Tree/Stack/94.binary-tree-inorder-traversal.cpp
/*
 * @lc app=leetcode id=94 lang=cpp
 *
 * [94] Binary Tree Inorder Traversal
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> out;
        stack<TreeNode*> stack;
        TreeNode* temp = root;
        while(!stack.empty()||temp){
            while(temp){
                stack.push(temp);
                temp = temp ->left;
            }
            temp = stack.top();
            stack.pop();
            out.push_back(temp->val);
            temp=temp->right;
        }
        return out;
    }
};

File: Tree/Recursion/94.binary-tree-inorder-traversal.java
/*
 * @lc app=leetcode id=94 lang=java
 *
 * [94] Binary Tree Inorder Traversal
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> out = new ArrayList<>();
        preTraversal(root, out);
        return out;
    }
    public void preTraversal(TreeNode root, List<Integer> out){
        if(root!=null){
            preTraversal(root.left, out);
            out.add(root.val);
            preTraversal(root.right, out);
        }
        else return;
    }
}

